[blog] Add 0.73 announcement post (#3931)
facebook/react-native-website: website/blog/2023-12-04-0.73-debugging-improvements-stable-symlinks.md

- [Kotlin Template on Android](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#kotlin-template-on-android)
- [Android 14 Support](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#android-14-support)
- [Deprecated Debugging Features](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#deprecated-debugging-features)
- [New Architecture Updates](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#new-architecture-updates)

### Breaking Changes

- [Babel Package Renames](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#babel-package-renames)
- [React Native CLI Changes](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#react-native-cli-changes)
- [Other Breaking Changes](/blog/2023/12/04/0.73-debugging-improvements-stable-symlinks#other-breaking-changes)

<!--truncate-->

## Highlights

### Debugging Improvements

The React Native and Hermes teams are committed to improving the debugging experience in React Native. In 0.73, we're happy to share some initial progress out of this ongoing investment.

> #### Console Log History in Hermes

Author: 1/ Across reloads, no — should clear as per browsers. Across Fast Refresh pushes, we can't claim that reliably right now.
2/ It is. I've added a clarifying sentence.

`console.log()` is an ever popular way for developers to quickly debug their JavaScript code. In previous releases, console logs in React Native apps would not be recorded until a debugger was connected. This made it hard to observe logs that occur early during app load.

In React Native 0.73 we've addressed this issue. Hermes now captures all `console.log()` calls in the background, which will be sent to the Console tab when a debugger is first connected — matching the debugging experience in web browsers.

#### Updated Debugging Docs

We've refreshed the [Debugging](/docs/debugging) section of our docs, which now includes up-to-date information on how to connect all supported debuggers, more info on React DevTools, and refreshed visuals.

![Overview of the Debugging docs as of 0.73](/blog/assets/0.73-debugging-docs.jpg)

#### Experimental New Debugger

The React Native team is working on a new JavaScript debugger experience, intended to replace Flipper, with a Technical Preview available as of React Native 0.73. The new debugger opens immediately, and features a stripped-back Chrome DevTools UI customized for debugging React Native with Hermes.

:::note

> The new debugger is **experimental** and has some [known issues](https://github.com/react-native-community/discussions-and-proposals/discussions/733) we are actively working to solve in a future release of React Native.

Reviewer: ```suggestion
As of today, the new debugger is considered **experimental**. While you evaluate it, check the [known issues](https://github.com/react-native-community/discussions-and-proposals/discussions/733) that we are actively working on and report your feedback.
```

@huntie where should folks report feedback also?

:::

> ![The new debugger frontend opened to the "Welcome" pane](/docs/assets/debugging-debugger-welcome.jpg)

Reviewer: Is this image missing in this PR?

Author: Nope, part of the current docs :)

> Learn more about enabling this experience [in the docs](/docs/next/debugging?js-debugger=new-debugger#opening-the-debugger).

Author: I think we can skip this(?). Only because I think there's reasonable delta on the debugging parts (timeline-wise) that don't need re-emphasising right here — in addition to the talk closely matching content that is provided directly in this post.

It might make sense to link the RN EU talk against the New Architecture section. Will pursue 👍🏻

### Stable Symlink Support in Metro

> Support for resolving symlinks in Metro is now enabled by default. Symlink support enables React Native to work with monorepo setups when containing directories are configured with [`watchFolders`](https://metrobundler.dev/docs/configuration/#watchfolders).

Reviewer: nit: I'll add some bold here

Symlinks are deeply represented in Metro's internals, meaning they work with features such as [Fast Refresh](/docs/fast-refresh), and incur little performance overhead during bundling. Symlinks are supported on all desktop platforms, with and without Watchman.

:::info
#### Monorepo workarounds

We are aware there are still edge cases when using React Native in a monorepo layout. We have planned work to address some of these, which didn't quite make it into 0.73 — but which we aim to ship as soon as possible.

**For React Native template projects (`npx react-native init`)**, you will need to configure any [`watchFolders`](https://metrobundler.dev/docs/configuration/#watchfolders) outside of the project root in order for Metro to discover them ([more info](https://metrobundler.dev/docs/configuration/#unstable_enablesymlinks-experimental)). You may also need to update file paths if your `react-native` dependency is installed to a folder at a different level.

**For Expo apps**, support for Yarn (Classic) workspaces is configured out of the box. See also the [Work with monorepos](https://docs.expo.dev/guides/monorepos/) guide in the Expo docs.

> :::

Reviewer: IMHO this should be moved to a separate page or article that we should link here

Author: I agree it is an awkward block of text, but at the same time having this information inline is:

- Honest and up front about remaining limitations.
- Re-commits to shipping the remaining work soon (which is both looking very tangible (cc @robhogan), and follows RN EU talk guidance).
- Justified to include given we have named Stable Symlink Support as a headlining feature.
    - If we want to reconsider this, I'm open to it!

### Kotlin Template on Android

We're excited to announce that, starting from 0.73, Kotlin is now the **recommended language** for Android apps built with React Native. This follows the direction that the Android ecosystem has been moving in for several years and allows you to write your app using a modern language.

We've updated React Native's template on Android to use Kotlin instead of Java. The new `MainActivity.kt` and `MainApplication.kt` files are 36% smaller in size.

> The upgrade helper has also been updated to make it easier to migrate your `.java` file to `.kt` files.

Reviewer: ```suggestion
The [upgrade helper](https://react-native-community.github.io/upgrade-helper/) has also been updated to make it easier to migrate your `.java` file to `.kt` files.

If you updated the Java files in your project and you need support migrating them to Kotlin, you can use the `Code > Convert Java file to Kotlin File` utility of Android Studio (also accessible with the <key>⌘</key> + <key>⇧</key> + <key>⌥</key> + <key>K</key> keyboard shortcut).
```

@TheSavior I guess that's the best we can recommend.

@huntie if <key> doesn't render correctly on the website, feel free to remove it

### Android 14 Support

> We've updated some capabilities to support Android 14.

Reviewer: ```suggestion
We've updated React Native to fully support Android 14. Starting from 0.73, React Native developers can now target the latest Android SDK version, [API Level 34](https://developer.android.com/guide/topics/manifest/uses-sdk-element?hl=en#ApiLevels) (*Upside Down Cake*).
```

#### Grant partial access to photos and videos

[Selected Photos Access](https://developer.android.com/about/versions/14/changes/partial-photo-video-access) allows Android 14 users to grant apps access to specific items in their media library, rather than access to all media. In 0.73, React Native apps now support this capability, by using the `READ_MEDIA_VISUAL_USER_SELECTED` permission in the [`PermissionsAndroid`](https://reactnative.dev/docs/permissionsandroid) API.

#### Context-registered broadcasts

> When using SDK >= 34, we've migrated [Broadcast Receivers](https://developer.android.com/guide/components/activities/process-lifecycle) to support context-registered broadcasts — part of the app lifecycle model — in Android 14. This means React Native apps will behave predictably on Android 14 devices.

Reviewer: This doesn't really makes sense how it's written now (@huntie @lunaleaps). I've rewritten it but it's also a really niche change so not sure if we need it mentioned.

```suggestion
Android 13 introduced a flag to specify if dynamically registered [Broadcast Receivers](https://developer.android.com/guide/components/activities/process-lifecycle) should be exported or not. This helped to increase security of apps that are using Broadcast Receivers registered dynamically via `Context`. With Android 14, this flag is now mandatory and every dynamically-registered Broadcast Receiver must specify it.
We've updated React Native to support this API so that you can keep on using Broadcast Receivers also on the latest version of Android.
```

#### Target SDK version

> In 0.73, React Native developers can now target the latest Android SDK version, [API Level 34](https://developer.android.com/guide/topics/manifest/uses-sdk-element?hl=en#ApiLevels) (*Upside Down Cake*).

Reviewer: Let's remvoe this as I merged it with the intro paragrahp

Author: Haha I had the same thought overnight 👍🏻

> 

Reviewer: ```suggestion
#### Min SDK Bump

React Native 0.73 will be the latest version to support Android 5.0 (API Level 21). The upcoming version of React Native will have a min SDK of 23 (Android 6.0). You can read more about the upcoming changes to min SDK bump [over here](TODO). 
```

cc @huntie we need to add a link here

Author: @cortinico Do you know what this should link to or is likely to be?

#### Remote JavaScript Debugging

Remote JavaScript Debugging is a legacy debugging mode that connects an external web browser (Chrome) to your app and runs your JavaScript code inside a web page, i.e. `http://localhost:8081/debugger-ui`. This model could lead to inconsistent app behaviour while debugging, and is incompatible with native modules under the New Architecture.

In 0.73, Remote JavaScript Debugging is deprecated and has been removed from the Dev Menu. Enabling the remote debugger must now be done manually via the `NativeDevSettings` API. Doing this is covered in the [Other Debugging Methods docs](/docs/next/other-debugging-methods#remote-js-debugging).

:::info
Remote JavaScript Debugging was previously the default debugging experience for apps using JavaScriptCore (JSC). We recommend [Safari Developer Tools (direct JSC debugging)](/docs/next/other-debugging-methods#safari-developer-tools-direct-jsc-debugging) instead, for iOS apps.

We recommend using [Hermes](/docs/hermes) for a consistent debugging experience on all platforms.
:::

### New Architecture Updates

> We continue active development on the New Architecture and are fully invested in bringing it to the OSS community. You can read the latest updates in the [New Architecture working group](https://github.com/reactwg/react-native-new-architecture/discussions).

Reviewer 2: I know we don't want to talk much about this in the blog post, but I think it would be helpful if we could say more. I think people have questions, and looking at the link doesn't really talk to that progress or answer any questions they might have

Reviewer: @huntie agree. We need 2x paragraphs here:
1. One on Bridgeless/SVC being available for experimentation
2. One on Native Module interop
And they should link to the post we'll be posting in the New Architecture WG

Author: I've done a first pass at this, with some input from @dmytrorykun.

## Breaking Changes

### Babel Package Renames

> We've relocated two Babel-related packages out of Metro and into React Native's repository and versioning scheme. The new versions of these packages support New Architecture features in 0.73, meaning these dependencies must be updated.

Reviewer 2: Can you include a phrase or sentence on "why"? What was the improvement we were after that necessitated this change?

> Please follow the [upgrade helper](https://react-native-community.github.io/upgrade-helper/) when upgrading, to ensure you have updated these dependencies. If you were using `metro-react-native-babel-preset` in your Babel config file, don't forget to update that too.

Reviewer 2: > don't forget to update that too.

Reading this i'd think "Update, how?!" and then have to read the next section. 

Something like:

> Please follow the [upgrade helper](https://react-native-community.github.io/upgrade-helper/) when upgrading, to ensure you have updated these dependencies. If you were using `metro-react-native-babel-preset` in your Babel config file, you'll need to use the new package name.

or

> Please follow the [upgrade helper](https://react-native-community.github.io/upgrade-helper/) when upgrading, to ensure you have updated these dependencies. Some packages have been renamed:

> There are no changes to the [react-native](https://www.npmjs.com/package/react-native) package.

Reviewer 2: This is confusing. There are lots of changes within the package, and you need to bump the version. Is this saying that it hasn't been renamed?

Reviewer: +1 here. Let's just remove this sentence?

| Old Package Name                       | New Package Name                        |
| -------------------------------------- | --------------------------------------- |
| `metro-react-native-babel-preset`      | `@react-native/babel-preset`            |
| `metro-react-native-babel-transformer` | `@react-native/metro-babel-transformer` |

:::info
`@react-native/babel-preset` now includes `@react-native/babel-plugin-codegen`, this no longer needs to be specified separately in your Babel config file.
:::

### React Native CLI Changes

#### Highlighted breaking changes

> - Change default `taskPrefix` in build-android command from `assemble` to `bundle` ([#1913](https://github.com/react-native-community/cli/pull/1913))

Reviewer: Can we add a short sentence here explaining why this is happening? @szymonrybczak

Reviewer 3: ```suggestion
- Change default task prefix in `build-android` command. From now on, when you run `build-android`, the `bundle` task will be run instead of `assemble` ([#1913](https://github.com/react-native-community/cli/pull/1913))
```

> - Remove fallback flow for Metro config defaults ([#1972](https://github.com/react-native-community/cli/pull/1972))

Reviewer 2: This sounds scary, and the message or link doesn't tell the user what they should be doing instead

Reviewer: +1

- Remove `--configuration` option from `run-ios` (replaced with `--mode`) ([#2028](https://github.com/react-native-community/cli/pull/2028))
- Remove `--variant` option from `build-android` command (replaced with `--mode`) ([#2026](https://github.com/react-native-community/cli/pull/2026)).

[See full changelog for v12.0.0](https://github.com/react-native-community/cli/releases/tag/v12.0.0).

### Other Breaking Changes

> - Raise minimum Node.js requirement to 18.x ([#37709](https://github.com/facebook/react-native/pull/37709))

Reviewer 2: I think of breaking changes as being scary to users, so you should think about how to be specific to not unnecessary scare the reader. Sometimes that means including a why (like saying 'due to Node 16 EOL', or more nuance, like Nick's point on TypeScript.

That's how I think about the Metro fallback comment above too.

> - Major bump of TypeScript to 5.0 ([#36862](https://github.com/facebook/react-native/pull/36862))

Reviewer 4: A nuance here is that we explicitly allow the types to be used down to 4.8, but the template comes with 5.0 for typechecking improvements with ESModule resolution.

Reviewer 2: Great point, so something like

"Template now uses TypeScript 5.0. React Native types continue working on TypeScript 4.8"

- **Android**: Major bump of Fresco to 3.0 ([#38275](https://github.com/facebook/react-native/pull/38275))
- **iOS**: Raise minimum iOS version to 13.4 ([#36795](https://github.com/facebook/react-native/pull/36795))
- **iOS**: Metro will no longer be automatically started when running builds via Xcode ([#38242](https://github.com/facebook/react-native/pull/38242))

For library authors:

> - **Android**: Bump to AGP 8.0 ([discussion](https://github.com/react-native-community/discussions-and-proposals/issues/671))

Reviewer: ```suggestion
- **Android**: Bump to AGP 8.1.1 ([discussion](https://github.com/react-native-community/discussions-and-proposals/issues/671))
```

## Acknowledgements

React Native 0.73 contains over [2259 commits](https://github.com/facebook/react-native/compare/v0.72.7...v0.73.0) from 68 contributors. Thanks for all your hard work!

## Upgrade to 0.73

> Please use the [React Native Upgrade Helper
> ](https://react-native-community.github.io/upgrade-helper/) to view code changes between React Native versions for existing projects, in addition to the [Upgrading docs](/docs/upgrading). You can also create a new project with `npx react-native@latest init MyProject`.

Reviewer: ```suggestion
Please use the [React Native Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) to view code changes between React Native versions for existing projects, in addition to the [Upgrading docs](/docs/upgrading). You can also create a new project with `npx react-native@latest init MyProject`.
```
